backends-velox/src/main/scala/io/glutenproject/execution/RowToVeloxColumnarExec.scala

```diff
@@ -120,6 +120,9 @@ case class RowToVeloxColumnarExec(child: SparkPlan)
               val estimatedBufSize = Math.max(
                 Math.min(sizeInBytes.toDouble * numRows * 1.2, 31760L * numRows),
                 sizeInBytes.toDouble * 10)
+              // Here we limit the max buffer size to Int.MaxValue, which is the memory limit for
+              // UnsafeProject.
+              val estimatedNumRow = Math.min(Int.MaxValue / estimatedBufSize, numRows)
```

Comment:
Curious, after apply row to columnar, why do we still have `UnsafeProject`, should it be `ProjectExecTransformer` ?

Reply:
Here we fix the memory used problem while converting `UnsafeProject` to `Columnar`.  If the row data is largeI, [iterator.next()](https://github.com/oap-project/gluten/blob/main/backends-velox/src/main/scala/io/glutenproject/execution/RowToVeloxColumnarExec.scala#L135) would failed with `Cannot grow BufferHolder by size 1602585576 because the size after growing exceeds size limitation 2147483632`.
More details can be found in #3034.

Reply:
It's a good question, I would check the codegen to find out why `writer.reset()` is not called.